ateam_ai/src/trajectory_generation/ilqr_problem.hpp

```diff
@@ -0,0 +1,290 @@
+
+    return out;
+  }
+
+  Trajectory calculate(const State & initial_state)
+  {
+    forward_rollout(initial_state);
+    backward_pass();
+
+    return trajectory;
+  }
+
+private:
+  using Jacobians = std::array<Jacobian, T>;
+  using Hessians = std::array<Hessian, T>;
+  using Gradiants = std::array<Gradiant, T>;
+
+
+  void forward_rollout(const State & initial_state)
+  {
+    // Step 1: Forward Rollout
+    trajectory.front() = initial_state;
+    actions.front().setZero();
+    overall_cost = cost(trajectory.at(0), actions.at(0), 0);
+
+    for (Time t = 1; t < T; t++) {
+      trajectory.at(t) = dynamics(trajectory.at(t - 1), actions.at(t - 1), t);
+      actions.at(t).setZero();
+      overall_cost += cost(trajectory.at(t), actions.at(t), t);
+    }
+
+    // Step 2: Calculate jacobian/hessian for x,u
+    for (Time t = 0; t < T; t++) {
+      j.at(t) = jacobian(trajectory.at(t), actions.at(t), t);
+
+      g.at(t) = gradiant(trajectory.at(t), actions.at(t), t);
+      h.at(t) = hessian(trajectory.at(t), actions.at(t), t, g.at(t));
+    }
+  }
+
+  void backward_pass()
+  {
+    std::size_t num_iterations = 0;
+
+    while (num_iterations < max_num_iterations) {
+      // Step 3: Determine best control signal update
+      Feedforwards k;
+      Feedbacks K;
+
+      Gradiant Q_dxu;  // Derivative of V(f(x, u))
+      Hessian Q_ddxu;
+      State v_dx = g.back().block(0, 0, X, 1);  // Derivative of V(x)
+      Eigen::Matrix<double, X, X> v_ddx = h.back().block(0, 0, X, X);
+
+      k.at(T - 1).setZero();
+      K.at(T - 1).setZero();
+
+      for (Time t = T - 2; t >= 0; t--) {
+        // eq 4a
+        Q_dxu.block(0, 0, X, 1) =
+          g.at(t).block(0, 0, X, 1) +
+          j.at(t).block(0, 0, X, X).transpose() *
+          v_dx.block(0, 0, X, 1);
+        // eq 4b
+        Q_dxu.block(X, 0, U, 1) =
+          g.at(t).block(X, 0, U, 1) +
+          j.at(t).block(0, X, X, U).transpose() *
+          v_dx.block(0, 0, X, 1);
+        // eq 4c
+        Q_ddxu.block(0, 0, X, X) =
+          h.at(t).block(0, 0, X, X) +
+          j.at(t).block(0, 0, X, X).transpose() *
+          v_ddx.block(0, 0, X, X) *
+          j.at(t).block(0, 0, X, X);
+        // eq 4d
+        Q_ddxu.block(X, 0, U, X) =
+          h.at(t).block(X, 0, U, X) +
+          j.at(t).block(0, X, X, U).transpose() *
+          v_ddx.block(0, 0, X, X) *
+          j.at(t).block(0, 0, X, X);
+        // eq 4e
+        Q_ddxu.block(X, X, U, U) =
+          h.at(t).block(X, X, U, U) +
+          j.at(t).block(0, X, X, U).transpose() *
+          v_ddx.block(0, 0, X, X) *
+          j.at(t).block(0, X, X, U);
+
+        // eq 5b
+        // Solve inverse with regulization to keep it from exploding
+        Eigen::SelfAdjointEigenSolver<Eigen::Matrix<double, U, U>> eigensolver(Q_ddxu.block(
+            X, X, U,
+            U));
+        if (eigensolver.info() != Eigen::Success) {
+          std::cout << "Failed to solve eigen vectors" << std::endl;
+          abort();
```

Comment:
Consider throwing an exception instead so the node that ends up wrapping this can correctly log and maybe recover?

Reply:
Oops, this was just for debugging. For the short term, I might make the return a bool signaling success instead of doing exceptions.

In the long term, I would like to integrate `absl::Status` into our code base. I've really enjoyed our custom version of this at work and it makes investigating crashes super simple.